rcl/src/rcl/wait.c

```diff
@@ -79,6 +80,9 @@ __wait_set_clean_up(rcl_wait_set_t * wait_set, rcl_allocator_t allocator)
     assert(ret == RCL_RET_OK);  // Defensive, shouldn't fail with size 0.
   }
   if (wait_set->impl) {
+    if (wait_set->impl->rmw_waitset) {
+      assert(rmw_destroy_waitset(wait_set->impl->rmw_waitset) == RMW_RET_OK);
```

Comment:
I actually don't see any other problematic uses of asserts (where needed code is avoided with `NDEBUG`) right now.

Perhaps we need a custom assert (it's a fairly common thing in projects) which is always checked. I think there's a need for "I want to make a defensive check that should never be false, and if it is it is not recoverable" but that is not disabled when `-DNDEBUG` is used. The consensus on the internet is that you can easily make one with `printf` and `abort` which are portable.